text/0003-data-frames.md

```diff
@@ -0,0 +1,252 @@
+
+Rather than operating on a single row, commands will need to be updated to handle a frame at a time. Here, the mapping should largely be the same, though an additional inner loop to process over the rows will be necessary. This processing can be done serial or in parallel and may be done synchronously or asynchronously.
+
+Commands that worked over inner tables should be able to migrate to data frames, as there is a strong overlap in functionality.
+
+Commands that filter will work similarly as before, and may opt to output streams which are flattened by the output stream. This allows them to optionally return no frame if no rows in the frame met the requirements of the filter.
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+Some drawbacks come to mind:
+
+- This is a large, non-trivial amount of work. Getting this landed, updating the commands to use the new model, and thoroughly testing will take time.
+- This will break most, if not all, third-party plugins
+
+# Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Everything is a frame
+
+One alternative is to require everything to live inside of a frame. There are some advantages here: this is seemingly more uniform, but at the risk of overloading the data frame concept.
+
+# Prior art
+
+[prior-art]: #prior-art
+
+Other data processing systems and languages have a data frame concept. The R language and the `pandas` library for Python use it as a model for working with data in tabular format.
+
+## R data frame
+
+```r
+Live Demo
+
+# Create the data frame.
+emp.data <- data.frame(
+   emp_id = c (1:5),
+   emp_name = c("Rick","Dan","Michelle","Ryan","Gary"),
+   salary = c(623.3,515.2,611.0,729.0,843.25),
+
+   start_date = as.Date(c("2012-01-01", "2013-09-23", "2014-11-15", "2014-05-11",
+      "2015-03-27")),
+   stringsAsFactors = FALSE
+)
+# Print the data frame.
+print(emp.data)
+```
+
+which outputs:
+
+```
+ emp_id    emp_name     salary     start_date
+1     1     Rick        623.30     2012-01-01
+2     2     Dan         515.20     2013-09-23
+3     3     Michelle    611.00     2014-11-15
+4     4     Ryan        729.00     2014-05-11
+5     5     Gary        843.25     2015-03-27
+```
+
+## Pandas data frame
+
+Below is an example of the pandas data frame:
+
+```python
+>>> d = {'col1': [1, 2], 'col2': [3, 4]}
+>>> df = pd.DataFrame(data=d)
+>>> df
+   col1  col2
+0     1     3
+1     2     4
+```
+
+# Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+- Are there syntactic ambiguities with the proposed syntax? This will require that we support parsing data frames, which includes colons and commas at the end of bare words.
+- How do we want to handle partial inner data frames? That is, a data frame that is inside of another data frame.
+- How do we handle non-data frames in between data frames? Do all partial data frames have to stream out until complete?
+
+# Future possibilities
```

Comment:
Maybe I missed it above but does anything live outside of DataFrame?

Reply:
For this proposal, yes, things can live outside of DataFrames. DataFrames are the way of creating data structures with primitive data, but you can also stream primitive data by itself.